README.md
# e621 Python Downloader
 A simple python script to download e621 images with specified tags and image amounts - Works with new API
## Branches
 The dev branch is for development builds, which may be buggy or entirely broken. Use these if you want to help with bug fixing.
## Releases
 I have built the python files into an .exe to prevent confusion, just download, unzip and run!
 Note: There isnt a release for V5 because pyinstaller is kinda broken lol


changelog.md
## V1.0:
* First Release:
* Very unoptimised, will do a whole rewrite soon.
## V2.0:
* Full Rewrite:
* Now much quicker to download.
## V2.1:
* No longer re-downloads images.
* Minor bug fixes.
## V2.2:
* Fixed bug where it would download 75 images max.
## V3.5:
* Pictures now downloaded into tag folder
* Allowed downloading without API Key/Username
* Made error messages slightly more userfriendly.
## V3.51:
* Fixed bug where it would re-download images >.<
## V4:
* New GUI! Very basic at the moment but hoping to add more features in the future!
* P.S: The GUI locks up when its starts downloading, it will only update after each image has been downloaded.
## V4.01:
* Fixed bug where if no tags entered (ie all tags) it would hang.
## V4.1:
* Bunch of bug fixes, removed downloading without API Key/Username.
## V5:
* Changed spaces to tabs (like wtf spaces are so jank)
* Changed max file name length to 255 (Actual windows max)
* Reformatted the changelog
* (Mostly) Fixed GUI locking up during image download
* Added GUI download progress bar
* Updated code comments

e621 downloader.py
import requests
import time
import json
import os
ratings = str(input("Please enter what rating you want [(-)Safe, (-)Questionable, (-)Explicit, All]:\n"))
if ratings.lower() == "safe":
    rating = "rating:safe"
elif ratings.lower() == "questionable":
    rating = "rating:questionable"
elif ratings.lower() == "explicit":
    rating = "rating:explicit"
elif ratings.lower() == "-safe":
    rating = "-rating:safe"
elif ratings.lower() == "-questionable":
    rating = "-rating:questionable"
elif ratings.lower() == "-explicit":
    rating = "-rating:explicit"
else:
    rating = ""
tag = str(input("Please enter what tags you would like (separated with a semicolon):\n"))
tags = tag.split(";")
postnum = str(input("Please enter how many posts you would like to download (Beware, a large number will take a while to download):\n"))
# Get user-defined url
url = "https://e621.net/posts.json?tags=" # Generates first static part of e621.net api url
for x in tags: # Add all tags specified
    url += x
    url += "%20"
url += rating
url += "&limit={0}&callback=callback".format(postnum) # Add rest of url
req = requests.get(url) # Sends a get request to server to get back urls of images
data = req.json() # Records the response as json
data = str(data)
res = [i for i in range(len(data)) if data.startswith("https://static1.e621.net/data", i)] # Find all strings starting with the e621 url
x = 0
while x < len(res):
    startpos = res[x]
    endpos = res[x]+73 # Standard url length
    strings = data[startpos:endpos] # Set strings to whole url
    if "preview" in strings or "sample" in strings: # Find if it is a low-res image
        data = data [:startpos] + data[endpos+8:] # Remove the url
        res.pop(x)
    else:
        x += 1
    res = [i for i in range(len(data)) if data.startswith("https://static1.e621.net/data", i)] # Update res
urls = []
x = 0
while x < len(res):
    res = [i for i in range(len(data)) if data.startswith("https://static1.e621.net/data", i)] # Update res
    startpos = res[x]
    endpos = res[x]+73 # Standard url length
    tempurl = data[startpos:endpos]
    tempurl = tempurl.replace("'","")
    urls.append(tempurl)
    x += 1
x = 0
urls.sort()
while x < len(urls):
    try:
        if urls[x] == urls[x+1]:
            urls.pop(x+1)
    except:
        y = 0
    fileName = urls[x]
    fileName = fileName[36:]
    img = requests.get(urls[x])
    cwd = os.getcwd() # Get current directory
    filePath = cwd + "\\Downloads\\" + fileName # Set download folder
    with open(filePath,"wb") as image:
        print("Downloading {0}".format(fileName))
        image.write(img.content)
    x += 1
    time.sleep(1) # Rate Limit


New WIP Version/new e621 downloader.py
import requests
from requests.auth import HTTPBasicAuth
import time
import json
import os
import sys

ratings = str(input("Please enter what rating you want [(-)Safe, (-)Questionable, (-)Explicit, All]:\n"))
if ratings.lower() == "safe":
    rating = "rating:safe"
elif ratings.lower() == "questionable":
    rating = "rating:questionable"
elif ratings.lower() == "explicit":
    rating = "rating:explicit"
elif ratings.lower() == "-safe":
    rating = "-rating:safe"
elif ratings.lower() == "-questionable":
    rating = "-rating:questionable"
elif ratings.lower() == "-explicit":
    rating = "-rating:explicit"
else:
    rating = ""
    
tag = str(input("Please enter what tags you would like (separated with a semicolon):\n"))
tags = tag.split(";")

postnum = str(input("Please enter how many posts you would like to download (Beware, a large number will take a while to download):\n"))

defaultURL = "https://e621.net/posts.json"
currentFolder = os.cwd()
apiKeyFile = currentFolder + os.sep + "apikey.txt"



New_WIP_Version/new_e621_downloader.py
import requests
from requests.auth import HTTPBasicAuth
import time
import json
import os
import sys

defaultURL = "https://e621.net/posts.json"
currentFolder = os.path.dirname(os.path.realpath(__file__))
apiKeyFile = currentFolder + os.sep + "apikey.txt"

absoluteLimit = 320
rateLimit = 1
lastTime = time.time()
lowestID = 0
stop = False

headers = {"User-Agent":"E6-Post-Downloader/1.0 (by jezzar on E621 with help from pup)"}
params = {"tags":"protogen"}

def rateLimiting():
    global lastTime
    timeTaken = time.time() - lastTime
    if timeTaken <= rateLimit:
        time.sleep(rateLimit-timeTaken)
    lastTime = time.time()

try:
    with open(apiKeyFile) as apiFile:
        apiKeys = apiFile.read().splitlines()
except FileNotFoundError:
    with open(apiKeyFile,"w") as apiFile:
        apiFile.write("user="+os.linesep+"api_key=")
    print("apikey.txt created - Add your username and API Key (https://e621.net/users/home -> Manage API Access) then press enter to continue")
    input()
    exit()

apiUser = apiKeys[0].split("=")[1]
apiKey = apiKeys[2].split("=")[1]

if apiUser == "" or apiKey == "":
    print("Please actually enter your username and API Key (https://e621.net/users/home -> Manage API Access)...")
    input()
    exit()

ratings = str(input("Please enter what rating you want [(-)Safe, (-)Questionable, (-)Explicit, All]:\n"))
if ratings.lower() == "safe":
    rating = "rating:safe"
elif ratings.lower() == "questionable":
    rating = "rating:questionable"
elif ratings.lower() == "explicit":
    rating = "rating:explicit"
elif ratings.lower() == "-safe":
    rating = "-rating:safe"
elif ratings.lower() == "-questionable":
    rating = "-rating:questionable"
elif ratings.lower() == "-explicit":
    rating = "-rating:explicit"
else:
    rating = ""

tag = str(input("Please enter what tags you would like (separated with a semicolon):\n"))
tags = tag.split(";")

postnum = str(input("Please enter how many posts you would like to download (Beware, a large number will take a while to download):\n"))

req = requests.get(defaultURL, headers=headers, params=params, auth=HTTPBasicAuth(apiUser,apiKey))
data = req.json()

if req.status_code != 200:
    print(f"Couldn't contact server. Error code: {req.status_code}.")
    sys.exit()

while stop != True:
    if len(data["posts"]) < absoluteLimit:
        stop = True

    for posts in data["posts"]:
        postURL = posts["file"]["url"]
        fileName = postURL[36:]
        img = requests.get(postURL)
        cwd = os.path.dirname(os.path.realpath(__file__)) # Get current directory
        filePath = cwd + "\\Downloads\\" + fileName # Set download folder
        with open(filePath,"wb") as image:
            print(f"Downloading {fileName}")
            image.write(img.content)

    params = {"tags":""}

    if req.status_code != 200:
        print("Couldn't contact server.")


e621_downloader.py
import tkinter as tk
import requests, time, json, os, sys, hashlib, errno, glob, re, threading
from requests.auth import HTTPBasicAuth
#from tkinter.ttk import *
from tkinter.ttk import Combobox, Progressbar
from tkinter import HORIZONTAL
from threading import Thread

headers = {"User-Agent":"E6-Post-Downloader/4.1 (by jezzar on E621)"}

def rateLimiting():
	global lastTime, rateLimit
	timeTaken = time.time() - lastTime
	if timeTaken <= rateLimit:
		time.sleep(rateLimit-timeTaken)
	lastTime = time.time()

def startDownload():
	# Setup variables:
	global rateLimit, absoluteLimit, lastTime
	defaultURL = "https://e621.net/posts.json"
	pastURL = ""
	currentFolder = os.path.dirname(os.path.realpath(__file__))
	apiKeyFile = os.path.join(currentFolder, "apikey.txt")
	stop = False
	absoluteLimit = 320
	rateLimit = 1
	lastTime = time.time()
	lowestID = -1
	# Get API key and Username from apikey.txt
	try:
		with open(apiKeyFile) as apiFile:
			apiKeys = apiFile.read().splitlines()
	except FileNotFoundError:
		with open(apiKeyFile,"w") as apiFile:
			apiFile.write("user="+os.linesep+"api_key=")
		outputLabel.config(text="apikey.txt created - Add your username and API Key (https://e621.net/users/home -> Manage API Access) then press download again.")
		return
	GUI.update()
	apiUser = apiKeys[0].split("=")[1]
	apiKey = apiKeys[2].split("=")[1]

	if apiUser == "" or apiKey == "":
		outputLabel.config(text="Please actually put your username and API key into apikey.txt")
		return
	GUI.update()
	naming = namingList.get()
	if naming.lower() != "md5":
	# Set tag file naming default
		naming = "tags"
	# Set how many general tags to write
		maxtags=namingTagNumEntry.get()
		if maxtags == (""):
			maxtags = 5
	else:
		naming = "md5"
	# Set appropriate ratings
	ratings = ratingList.get()
	if ratings == "Safe":
		rating = "rating:safe"
	elif ratings == "Questionable":
		rating = "rating:questionable"
	elif ratings == "Explicit":
		rating = "rating:explicit"
	elif ratings == "-Safe":
		rating = "-rating:safe"
	elif ratings == "-Questionable":
		rating = "-rating:questionable"
	elif ratings == "-Explicit":
		rating = "-rating:explicit"
	else:
		rating = ""

	seg1 = tagEntry.get()
	if seg1 != (""):
		tag = str(seg1)
		tags = tag.split(" ")
	else:
		tags = ""

	grabURL = f"{defaultURL}?tags="
	x = " ".join(tags)
	grabURL += x
	grabURL += f" {rating}&limit=320"

	downfoldersw = downFolderEntry.get()
	if downfoldersw:
		downfolder = downfoldersw.capitalize()
	elif tags == "": # If no tags entered, rating is used as the folder name.
		downfolder = ratings.title().replace("-", "Not ")
	else: # If nothing is typed, tags will be used as folder name
		downfolder = tag.title()
	GUI.update()
	if not os.path.exists(os.path.join(currentFolder, "Downloads", downfolder)):
		try: # Create folder
			os.makedirs(os.path.join(currentFolder, "Downloads", downfolder))
			outputLabel.config(text=f"Folder {downfolder} created!\nStarting download!")
		except OSError as e:
			if e.errno != errno.EEXIST:
				raise
	else:# Start download if already exist
		outputLabel.config(text=f"Folder {downfolder} already exists!\nStarting download!")

	GUI.update()
	req = requests.get(grabURL, headers=headers, auth=HTTPBasicAuth(apiUser,apiKey))
	data = req.json()

	if req.status_code != 200:
		outputLabel.config(text=f"Couldn't contact e621. Error code: {req.status_code}.\nJson: {data}")
		sys.exit()

	downloaded = []
	for _, _, files in os.walk(currentFolder + os.path.sep + "Downloads"):
		for name in files:
			downloaded.append(name.split(".")[0])
	GUI.update()
	# Alot of checks for file naming (based on SkylerMews code)
	def safe_filename(string):
		# Set valid windows characters
		set = '-_.() abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'
		# Replace ♥ emotes
		string = string.replace("<3_eyes","heart_eyes")
		string = string.replace("<3_penis","heart_penis")
		string = string.replace("<3_tail","heart_tail")
		string = string.replace("<3_censor","heart_censor")
		# Remove ♥ emotes
		string = string.replace("<3","")
		string = string.replace("</3","")
		# Remove aspect ratio tags
		string = string.replace("1:2","")
		string = string.replace("2:3","")
		string = string.replace("3:4","")
		string = string.replace("4:5","")
		string = string.replace("5:6","")
		string = string.replace("1:1","")
		string = string.replace("5:4","")
		string = string.replace("4:3","")
		string = string.replace("3:2","")
		string = string.replace("16:10","")
		string = string.replace("7:4","")
		string = string.replace("16:9","")
		string = string.replace("2:1","")
		# Replace some
		string = string.replace("/","-")
		string = string.replace(":","-")
		# Remove elipsis
		string = string.replace("...","")
		# Remove multiple spacing
		string = ' '.join(string.split())
		s = ''
		for c in string:
			if set.find(c) != -1:
				s = s + c
		return s
	GUI.update()
	while stop != True:
		if len(data["posts"]) > absoluteLimit:
			stop = True
		GUI.update()
		for posts in data["posts"]:
			postURL = posts["file"]["url"]
			if lowestID > posts['id'] or lowestID == -1:
				lowestID = posts['id']
			if pastURL == postURL:
				continue
			ArtistTags = " ".join(posts["tags"]["artist"])+" "
			CharacterTags = " ".join(posts["tags"]["character"])+" "
			CopyrightTags = " ".join(posts["tags"]["copyright"])+" "
			GeneralTags = " ".join(posts["tags"]["general"])
			fileName = ArtistTags + CopyrightTags + CharacterTags
			if len(fileName) > 255:
					fileName = fileName[:255]
			if posts["file"]["md5"] in downloaded:
				outputLabel.config(text="Skipping existing post.")
				GUI.update()
				continue
			if fileName in downloaded:
				outputLabel.config(text="Skipping existing post.")
				GUI.update()
				continue
			if len(glob.glob(downfolder+"/"+str(lowestID)+' *')) > 0: # Checks if the image already exists
				 outputLabel.config(text=f'File already exists. Post ID: "{lowestID}"')
				 GUI.update()
				 continue
			pastURL = posts["file"]["url"]
			img = requests.get(postURL)
			if naming.lower() == "md5": # Sets correct name according to user selected option
				fileName = postURL[36:]
				filePath = os.path.join(currentFolder, "Downloads", downfolder, fileName)
			elif naming.lower() == "tags":
				if len(fileName) > 255:
					fileName = fileName[:255]
				fileExt = "." + pastURL.split(".")[3]
				writeName = str(lowestID) + " " + safe_filename(fileName+' '.join(GeneralTags.split()[:int(maxtags)]))
				fullName = writeName + fileExt
				filePath = os.path.join(currentFolder, "Downloads", downfolder, fullName)
			img = requests.get(postURL, stream=True)
			with open(filePath,"wb") as image: # Fancy progress bar code
				imageSize = int(img.headers["content-length"])
				chunkSize = imageSize // 100 # So it has a width of 100
				downloadBar = 0
				for chunk in img.iter_content(chunk_size=chunkSize):
					outputLabel.config(text=f"Downloading {fileName}")
					progressBar["value"] = downloadBar
					GUI.update()
					image.write(chunk)
					downloadBar += 1

		rateLimiting()

		grabURL = f"{defaultURL}?tags="
		x = " ".join(tags)
		grabURL += x
		grabURL += f" {rating}&page=b{lowestID}&limit=320" # Create new URL
		
		GUI.update()
		
		req = requests.get(grabURL, headers=headers, auth=HTTPBasicAuth(apiUser,apiKey))
		data = req.json() # Get all new posts from the new URL

		GUI.update()
		
		if req.status_code != 200: # Check if it was successful
			outputLabel.config(text=f"Couldn't contact e621. Error code: {req.status_code}.\nJson: {data}")
			GUI.update()
			sys.exit()

def downloadThread():
	DownloadThread = Thread(target=startDownload)
	DownloadThread.start()
	while DownloadThread.is_alive:
		GUI.update()

# v GUI shit thats really annoying seriously tkinter is so shit

GUI = tk.Tk()

namingLabel = tk.Label(GUI,text="Naming Scheme")

namingOptions = ("md5","tags")
namingList = Combobox(GUI,values=namingOptions)

namingTagNumLabel = tk.Label(GUI,text="If naming as tags:\nhow many general tags would you like?\n(Default=5)")

namingTagNumEntry = tk.Entry(GUI,text="")

ratingLabel = tk.Label(GUI,text="Select Rating")

ratingOptions = ("All","Safe","Questionable","Explicit","-Safe","-Questionable","-Explicit")
ratingList = Combobox(GUI,values=ratingOptions)

tagLabel = tk.Label(GUI,text="Enter Tags")

tagEntry = tk.Entry(GUI,text="")

downFolderLabel = tk.Label(GUI,text="Enter the Download Folder\n(Default=tags)")

downFolderEntry = tk.Entry(GUI,text="")

downloadButton = tk.Button(GUI,text="Download!",command=startDownload)

outputLabel = tk.Label(GUI,text="")

progressBar = Progressbar(GUI, orient=HORIZONTAL, length=200, mode="determinate")

namingLabel.grid(column=0,row=0)
namingList.grid(column=0,row=1)
namingTagNumLabel.grid(column=0,row=2)
namingTagNumEntry.grid(column=0,row=3)
ratingLabel.grid(column=0,row=4)
ratingList.grid(column=0,row=5)
tagLabel.grid(column=0,row=6)
tagEntry.grid(column=0,row=7)
downFolderLabel.grid(column=0,row=8)
downFolderEntry.grid(column=0,row=9)
downloadButton.grid(column=0,row=10)
outputLabel.grid(column=0,row=11)
progressBar.grid(column=0,row=12)

GUI.update()
GUI.mainloop()
